### ISSUE TITLE
Build times on CI

### ISSUE THREAD

**Issue by hs-apotell** [chipsalliance/Surelog#3322]
Build times on CI have gone up drastically, now almost 3 hours for each build. Is there a plan to pull back this to be under reasonable limits?

One proposal (not the only proposal) - Manually bucketing the regression tests into a handful of buckets where each bucket can be completed in reasonable time. The workflow will be more like a pipeline - run build and archive the tests as artifacts and each regression node will pull these artifacts and run the selected tests.

The obvious drawbacks -
* Bucketing is manual (adding a new test would require adding the same to one of these buckets)
* Results of regression are spread across multiple jobs (this can potentially be resolved by adding another step after all the regression runs are completed).

Any other ideas on how to cut back on the CI build times?

**Comment by hzeller** [comment]
Ok, so local sharding preferred to external runners (which indeed, I just asked, would only really work for Linux and require some work for Windows but will not be possible for Macs).

Anything other than hash would be either (a) manual (b) uneven (e.g. if you choose alphabetically).

So using the standard way of doing the sharding with hashes is the best of all. If you want predictability, this is why you'd use hashes as the same test always ends up in the same instance (unless you change number of instances, but that is rare).

Can you implement the suggested `--shard` feature in the Python regression @hs-apotell  then I can play around with making the CI work.

**Comment by proppy** [comment]
> @proppy you played with building things in stages: it should be possible to build the binary once and then use the output in a parallelized sharded job, right ? 

Yes https://github.com/actions/cache combined with https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs can definitly help for that.

@umarcor might also have some ideas!

**Comment by hzeller** [comment]
My current WIP approach is #3326 in which I store the built binaries to then be picked up by the next stage. Then sharding that next stage with a matrix.
